sdk/src/http/restricted.rs

```diff
@@ -0,0 +1,404 @@
+        }
+    }
+
+    /// Creates a new `RestrictedResolver` with the specified allowed list.
+    #[allow(dead_code)] // TODO: temp until http module is public
+    pub fn with_allowed_hosts(inner: T, allowed_hosts: Vec<HostPattern>) -> Self {
+        Self {
+            inner,
+            allowed_hosts: Some(allowed_hosts),
+        }
+    }
+
+    /// Replaces the current allowed list with the given allowed list if specified.
+    pub fn set_allowed_hosts(&mut self, allowed_hosts: Option<Vec<HostPattern>>) {
+        self.allowed_hosts = allowed_hosts;
+    }
+
+    /// Returns a reference to the allowed list.
+    pub fn allowed_hosts(&self) -> Option<&[HostPattern]> {
+        self.allowed_hosts.as_deref()
+    }
+}
+
+impl<T: SyncHttpResolver> SyncHttpResolver for RestrictedResolver<T> {
+    fn http_resolve(
+        &self,
+        request: Request<Vec<u8>>,
+    ) -> Result<Response<Box<dyn Read>>, HttpResolverError> {
+        if self
+            .allowed_hosts()
+            .is_none_or(|hosts| is_uri_allowed(hosts, request.uri()))
+        {
+            self.inner.http_resolve(request)
+        } else {
+            Err(HttpResolverError::UriDisallowed {
+                uri: request.uri().to_string(),
+            })
+        }
+    }
+}
+
+#[cfg_attr(not(target_arch = "wasm32"), async_trait)]
+#[cfg_attr(target_arch = "wasm32", async_trait(?Send))]
+impl<T: AsyncHttpResolver + Sync> AsyncHttpResolver for RestrictedResolver<T> {
+    async fn http_resolve_async(
+        &self,
+        request: Request<Vec<u8>>,
+    ) -> Result<Response<Box<dyn Read>>, HttpResolverError> {
+        if self
+            .allowed_hosts()
+            .is_none_or(|hosts| is_uri_allowed(hosts, request.uri()))
+        {
+            self.inner.http_resolve_async(request).await
+        } else {
+            Err(HttpResolverError::UriDisallowed {
+                uri: request.uri().to_string(),
+            })
+        }
+    }
+}
+
+/// A host/scheme pattern used to restrict network requests.
+///
+/// Each pattern may include:
+/// - A scheme (e.g. `https://` or `http://`)
+/// - A hostname, which may have a single leading wildcard (e.g. `*.contentauthenticity.org`)
+///
+/// Matching is case-insensitive. A wildcard pattern such as `*.contentauthenticity.org` matches
+/// `sub.contentauthenticity.org`, but does not match `contentauthenticity.org` or `fakecontentauthenticity.org`.
+/// If a scheme is present in the pattern, only URIs using the same scheme are considered a match. If the scheme
+/// is omitted, any scheme is allowed as long as the host matches.
+#[cfg_attr(
+    feature = "json_schema",
+    derive(schemars::JsonSchema),
+    schemars(with = "String")
+)]
+#[derive(Debug, Clone, PartialEq)]
+pub struct HostPattern {
+    pattern: String,
+    scheme: Option<String>,
+    host: Option<String>,
+}
+
+impl HostPattern {
+    /// Creates a new `HostPattern` with the given pattern.
+    pub fn new(pattern: &str) -> Self {
```

Comment:
Could you add a test with these patterns: 
`let pattern = HostPattern::new("https://"); `
and
`let pattern = HostPattern::new("https:// ");`
? 

Corollary - do we need port number handling too?

Reply:
> `let pattern = HostPattern::new("https:// ");`

What do you expect to happen in this case? I'd consider it undefined behavior because spaces are not allowed in URIs.